features/192-asset-gear/index.md

```diff
@@ -0,0 +1,667 @@
+# Asset Gear
+
+## Goals
+
+- Store user-generated asset to fulfill application needs.
+  - Photos and short videos taken by the application.
+  - Images and short videos chosen from the library.
+  - Artwork by the user.
+  - Attachments in a chat application.
+  - Documents of reasonable size.
+- Process image asset on the fly via query string.
+- Access control via presigned URL
+
+## Non-goals
+
+- Video live streaming such as Facebook Live.
+- File sharing application like Imgur.
+- Static website hosting. Another gear is responsible for this.
+
+## Layout
+
+The asset name is in the following format.
+
+```
+/<app-id>/<random-string>[.<ext>]
+```
+
+- `app-id`: The app ID.
+- `random-string`: A randomly generated string.
+- `ext`: The inferred file extension (if any) from the given content type.
+
+## HTTP API
+
+### POST /_asset/presign_upload
+
+#### Description
+
+This endpoint returns a presigned upload request which can be used to upload an asset.
+
+This endpoint requires API Key and authenticated user.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "exact_name": { "type": "string" },
+    "prefix": { "type": "string" },
+    "access": { "enum": ["public", "private"] },
+    "headers": {
+      "type": "object",
+      "properties": {
+        "content-type": { "type": "string" },
+        "content-disposition": { "type": "string" },
+        "content-encoding": { "type": "string" },
+        "cache-control": { "type": "string" },
+        "access-control-allow-origin": { "type": "string" },
+        "access-control-expose-headers": { "type": "string" },
+        "access-control-expose-headers": { "type": "string" },
+        "access-control-max-age": { "type": "string" },
+        "access-control-allow-credentials": { "type": "string" },
+        "access-control-allow-methods": { "type": "string" },
+        "access-control-allow-headers": { "type": "string" }
+      }
+    }
+  }
+}
+```
+
+- `exact_name`: The exact name of the asset.
+- `prefix`: If `exact_name` is not given, a random name is generated with `prefix` prepended.
+- `access`: The access control of the asset. `public` is the default.
+- `headers`: The HTTP headers of the asset when it is retrieved.
+- `headers.content-type`: The Content-Type header. `application/octet-stream` is the default.
+- `headers.content-disposition`: The Content-Disposition header.
+- `headers.content-encoding`: The Content-Encoding header. The developer must make sure the asset is compressed in the given compression.
+- `headers.cache-control`: The Cache-Control header. If it is not given, a default one is added. To remove the header, set it to an empty string.
+- `headers.access-control-*`: The CORS headers. They are only useful if the developer needs to retrieve the asset with Fetch API.
+
+##### Request Example
+
+```json
+{
+  "headers": {
+    "content-type": "image/png"
```

Comment:
why md5 and length is suddenly removed? I thought it was already explained and agreed it is optional for integrity checking?

Reply:
Whether or not Content-Length and Content-MD5 are required depends on the method of upload. If the server returns a PUT Object request, then both are required. If the server returns a POST Object request, then both are ignored.

I am going to add them back but they may be ignored in some case.